File: app01_java_nft/backend/src/main/java/com/devpowers/nftcerts/api/UploadStore.java
package com.devpowers.nftcerts.api;

import com.devpowers.nftcerts.error.FileNotFoundInUploadStoreException;
import org.springframework.stereotype.Component;

import java.util.Map;
import java.util.UUID;
import java.util.concurrent.ConcurrentHashMap;

/**
 * Simple in-memory store for uploaded (hashed/watermarked) file bytes, retrievable later by
 * {@code fileId} when creating an artwork. H2 is not used for blob storage per the design spec.
 */
@Component
public class UploadStore {

    private final Map<UUID, StoredUpload> uploads = new ConcurrentHashMap<>();

    public UUID store(StoredUpload upload) {
        UUID fileId = UUID.randomUUID();
        uploads.put(fileId, upload);
        return fileId;
    }

    public StoredUpload get(UUID fileId) {
        StoredUpload upload = uploads.get(fileId);
        if (upload == null) {
            throw new FileNotFoundInUploadStoreException(fileId);
        }
        return upload;
    }
}

File: app02_cpp_nft/backend/src/api/UploadStore.cpp
#include "api/UploadStore.h"

#include "error/Exceptions.h"
#include "util/Uuid.h"

namespace nftcerts::api {

std::string UploadStore::store(const certificate::StoredUpload& upload) {
    std::string fileId = util::randomUuid();
    std::lock_guard<std::mutex> lock(mutex_);
    uploads_[fileId] = upload;
    return fileId;
}

certificate::StoredUpload UploadStore::get(const std::string& fileId) {
    std::lock_guard<std::mutex> lock(mutex_);
    auto it = uploads_.find(fileId);
    if (it == uploads_.end()) {
        throw error::FileNotFoundInUploadStoreException(fileId);
    }
    return it->second;
}

}  // namespace nftcerts::api
